src/classes/SObjectRepository.cls

```diff
@@ -2,40 +2,98 @@
 
+    private final Schema.SObjectType sObjectType;
+    private final Schema.FieldSet fieldSet;
+    private final List<Schema.SObjectField> sobjectFieldList;
+    private final Schema.SObjectField idField;
     protected IQueryBuilder queryBuilder;
 
-    protected SObjectRepository(Schema.SObjectType sobjectType, Schema.FieldSet fieldSet) {
+    public SObjectRepository(Schema.SObjectType sobjectType, Schema.FieldSet fieldSet) {
         this(sobjectType, fieldSet, null);
     }
 
-    protected SObjectRepository(Schema.SObjectType sobjectType, List<Schema.SObjectField> sobjectFieldList) {
+    public SObjectRepository(Schema.SObjectType sobjectType, List<Schema.SObjectField> sobjectFieldList) {
         this(sobjectType, null, sobjectFieldList);
     }
 
     private SObjectRepository(Schema.SObjectType sobjectType, Schema.FieldSet fieldSet, List<Schema.SObjectField> sobjectFieldList) {
         this.currentModule = NebulaCore.Module.REPOSITORY;
 
-        this.queryBuilder = new QueryBuilder(sobjectType, fieldSet, sobjectFieldList);
+        this.sObjectType = sObjectType;
+        this.fieldSet = fieldSet;
+        this.sObjectFieldList = sObjectFieldList;
+        this.idField = this.getField(this.sObjectType, 'Id');
     }
 
-    public virtual void doInsert(SObject record) {this.doInsert(new List<SObject>{record});}
-    public virtual void doInsert(List<SObject> recordList) {Database.insert(recordList);}
-    public virtual void doUpdate(SObject record) {this.doUpdate(new List<SObject>{record});}
-    public virtual void doUpdate(List<SObject> recordList) {Database.update(recordList);}
-    // Apex doesn't allow upserting on generic SObject, so we'll delegate the implementation to each concrete class
-    public abstract void doUpsert(SObject record);
-    public abstract void doUpsert(List<SObject> recordList);
+    protected IQueryBuilder queryFactory {
+        get {
+            return new QueryBuilder(sObjectType, fieldSet, sObjectFieldList);
+        }
+    }
+
+    //SOQL
+    public virtual SObject getById(Id recordId) {
+        return this.queryFactory
+            .whereField(new List<QueryPredicate>{QueryPredicate.EQUALS(this.idField, recordId)})
+            .getFirstOrDefault();
+    }
+
+    public virtual List<SObject> getById(List<Id> recordIds) {
+        return this.queryFactory
+            .whereField(new List<QueryPredicate>{QueryPredicate.EQUALS(this.idField, recordIds)})
+            .getQueryResults();
+    }
+
+    public virtual List<SObject> get(QueryPredicate queryPredicate) {
+        return this.get(new List<QueryPredicate> { queryPredicate });
+    }
+
+    public virtual List<SObject> get(List<QueryPredicate> queryPredicates) {
+        return this.queryFactory
+            .whereField(queryPredicates)
+            .getQueryResults();
+    }
+
+    public virtual List<SObject> getByFieldForIds(Schema.SObjectField field, Object value, Set<Id> idSet) {
```

Comment:
lines 59 - 83 are the ones I am thinking seriously about replacing with query predicates in the calling implementations.